pkg/provisioner/longhornv1.go

```diff
@@ -0,0 +1,457 @@
+	nodeObj          *longhornv1.Node
+	nodesClientCache ctllonghornv1.NodeCache
+	nodesClient      ctllonghornv1.NodeClient
+
+	cacheDiskTags *DiskTags
+	semaphoreObj  *Semaphore
+}
+
+func NewLHV1Provisioner(
+	device *diskv1.BlockDevice,
+	block block.Info,
+	nodeObj *longhornv1.Node,
+	nodesClient ctllonghornv1.NodeClient,
+	nodesClientCache ctllonghornv1.NodeCache,
+	cacheDiskTags *DiskTags,
+	semaphore *Semaphore,
+) (Provisioner, error) {
+	baseProvisioner := &provisioner{
+		name:      TypeLonghornV1,
+		blockInfo: block,
+		device:    device,
+	}
+	provisioner := &LonghornV1Provisioner{
+		provisioner:      baseProvisioner,
+		nodeObj:          nodeObj,
+		nodesClient:      nodesClient,
+		nodesClientCache: nodesClientCache,
+		cacheDiskTags:    cacheDiskTags,
+		semaphoreObj:     semaphore,
+	}
+
+	if !cacheDiskTags.Initialized() {
+		return nil, errors.New(ErrorCacheDiskTagsNotInitialized)
+	}
+	return provisioner, nil
+}
+
+func (p *LonghornV1Provisioner) GetProvisionerName() string {
+	return p.name
+}
+
+func (p *LonghornV1Provisioner) Provision() (bool, error) {
+	logrus.Infof("%s provisioning Longhorn block device %s", p.name, p.device.Name)
+
+	nodeObjCpy := p.nodeObj.DeepCopy()
+	tags := []string{}
+	if p.device.Spec.Tags != nil {
+		tags = p.device.Spec.Tags
+	}
+	diskSpec := longhornv1.DiskSpec{
+		Type:              longhornv1.DiskTypeFilesystem,
+		Path:              extraDiskMountPoint(p.device),
+		AllowScheduling:   true,
+		EvictionRequested: false,
+		StorageReserved:   0,
+		Tags:              tags,
+	}
+
+	// checked the case that longhorn node updated but blockdevice CRD is not updated
+	synced := false
+	provisioned := false
+	if disk, found := p.nodeObj.Spec.Disks[p.device.Name]; found {
+		synced = reflect.DeepEqual(disk, diskSpec)
+		if !synced {
+			logrus.Warnf("The disk spec should not different between longhorn node and blockdevice CRD, disk: %+v, diskSpec: %+v", disk, diskSpec)
+		}
+	}
+	if !synced {
+		logrus.Debugf("Try to sync disk %s to longhorn node %s/%s again", p.device.Name, p.nodeObj.Namespace, p.nodeObj.Name)
+		nodeObjCpy.Spec.Disks[p.device.Name] = diskSpec
+		if _, err := p.nodesClient.Update(nodeObjCpy); err != nil {
+			return true, err
+		}
+		provisioned = true
+	}
+
+	if (synced && !diskv1.DiskAddedToNode.IsTrue(p.device)) || provisioned {
+		logrus.Debugf("Set blockdevice CRD (%v) to provisioned", p.device)
+		msg := fmt.Sprintf("Added disk %s to longhorn node `%s` as an additional disk", p.device.Name, p.nodeObj.Name)
+		setCondDiskAddedToNodeTrue(p.device, msg, diskv1.ProvisionPhaseProvisioned)
+	}
+
+	p.cacheDiskTags.UpdateDiskTags(p.device.Name, p.device.Spec.Tags)
+	return false, nil
+}
+
+func (p *LonghornV1Provisioner) UnProvision() (bool, error) {
+	logrus.Infof("%s unprovisioning Longhorn block device %s", p.name, p.device.Name)
+
+	// inner functions
+	updateProvisionPhaseUnprovisioned := func() {
+		msg := fmt.Sprintf("Disk not in longhorn node `%s`", p.nodeObj.Name)
```

Comment:
`Longhorn` is a product name, thus it should be title-case. Every occurrence across all files should be adapted.
```suggestion
		msg := fmt.Sprintf("Disk not in Longhorn node `%s`", p.nodeObj.Name)
```

Reply:
done